server/linkpreview.go

```diff
@@ -0,0 +1,165 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"golang.org/x/net/html"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+type linkPreview struct {
+	Title       string `json:"title,omitempty"`
+	Description string `json:"description,omitempty"`
+	ImageURL    string `json:"image_url,omitempty"`
+}
+
+var client = &http.Client{
+	Timeout: time.Second * 2,
+	CheckRedirect: func(req *http.Request, via []*http.Request) error {
+		if err := validateURL(req.URL.String()); err != nil {
+			return err
+		}
+		return nil
+	},
+}
+
+// previewLink handles the HTTP request, fetches the URL, and returns the link preview.
+func previewLink(w http.ResponseWriter, r *http.Request) {
+	// check authorization
+	uid, challenge, err := authHttpRequest(r)
+	if err != nil {
+		http.Error(w, "invalid auth secret", http.StatusBadRequest)
+		return
+	}
+	if challenge != nil {
```

Comment:
Just return an error like `http.StatusUnauthorized`. There is no need to support it here.